Plugins/NativeServerPlugin/src/directx_buffer_capturer.cpp

```diff
@@ -0,0 +1,258 @@
+#include "pch.h"
+
+#include "directx_buffer_capturer.h"
+#include "plugindefs.h"
+
+#include "webrtc/modules/video_coding/codecs/h264/h264_encoder_impl.h"
+
+using namespace Microsoft::WRL;
+using namespace StreamingToolkit;
+
+DirectXBufferCapturer::DirectXBufferCapturer(ID3D11Device* d3d_device) :
+	d3d_device_(d3d_device)
+{
+}
+
+void DirectXBufferCapturer::Initialize(bool headless, int width, int height)
+{
+	// Gets the device context.
+	d3d_device_->GetImmediateContext(&d3d_context_);
+
+#ifdef MULTITHREAD_PROTECTION
+	// Enables multithread protection.
+	ID3D11Multithread* multithread;
+	d3d_device_->QueryInterface(IID_PPV_ARGS(&multithread));
+	multithread->SetMultithreadProtected(true);
+	multithread->Release();
+#endif // MULTITHREAD_PROTECTION
+
+	// Initializes NVIDIA encoder.
+	webrtc::H264EncoderImpl::SetDevice(d3d_device_.Get());
+	webrtc::H264EncoderImpl::SetContext(d3d_context_.Get());
+
+	// Headless mode initialization.
+	headless_ = headless;
+	if (headless_)
+	{
+		ResizeRenderTexture(width, height);
+	}
+}
+
+void DirectXBufferCapturer::SendFrame(int64_t prediction_time_stamp)
+{
+	if (!headless_)
+	{
+		return;
+	}
+
+	SendFrame(render_texture_.Get(), prediction_time_stamp);
+}
+
+void DirectXBufferCapturer::SendFrame(ID3D11Texture2D* frame_buffer, int64_t prediction_time_stamp)
+{
+	if (!running_)
+	{
+		return;
```

Comment:
if this is code that we'd expect another developer to call i'd much prefer to see some error condition occur here instead of an early return...since it seems the caller should not call this signature unless they are running. that way the dev can better diagnose what's going on

Reply:
This is different with the case above. There may be some situations which the video capturer hasn't started (webrtc is still being initialized) before the first frame finishes rendering (SendFrame is called) so it just quietly ignores instead of throwing an error.

Reply:
i see, that makes sense - it _might_ be worth dropping a comment to that effect in here.